### PR TITLE
Object Detection (Mallet)

### PR DESCRIPTION
**Pull request by frawgii**
### **Description**
This PR refactors and adjusts several constants related to camera recording, bounding box tracking, ArUco tag detection, object detection, the Stanley controller, and general algorithm parameters. The primary goal is to fine-tune these settings for improved performance, reliability, and accuracy in the autonomy system. Notably, this PR introduces object detection for the mallet, a new bounding box tracker type (CSRT for general bounding boxes, KCF for ArUco tags), adjusts filtering parameters for Torch-based object detection, modifies Stanley controller gains and prediction parameters, and makes minor adjustments to stuck detection and approach states.

### **Changes Made**
- Fixed issue with Stanley controller by updating the cross-track control gain, maximum steering angle limit, prediction horizon, and prediction time step variables. Other minor adjustments were also made such as to the stuck state by reducing the minimum angular velocity threshold, as well as modifying the camera recording toggles to all be explicitly defined as true.
- Introduced new feature for **mallet detection**. Though the trained model still needs to be added, logic for approaching and verifying the mallet is in the codebase. Handlers, states, and autonomy constants were all either created or updated to account for this new feature. The rover should be able to successfully approach a given object and verify that it has reached the mallet by flashing green, similarly to how it does for a tag.
- Improved performance of the bounding box and ArUco detection by updating and modifying their constants.

### **How It Works**
**1. Camera Recording:** The boolean constants now explicitly enable or disable recording for each camera source and its associated tag/object detectors. This provides clear control over which data streams are saved.

**2. Bounding Box Tracking:** The introduction of BBOX_TRACKER_IOU_MATCH_THRESHOLD for both general and ArUco bounding boxes allows the system to more intelligently associate new detections with existing tracked objects based on the Intersection over Union (IoU) of their bounding boxes. Using CSRT for general tracking and KCF for ArUco tags leverages the strengths of each algorithm for potentially more robust tracking.

**3. Object Detection (Torch)**: Reducing the data retrieval threads might optimize resource usage. Lowering the FPS for the main object detector could improve performance by reducing processing load. Increasing the confidence threshold and adjusting the NMS threshold for the Torch model aims to filter out less reliable detections and reduce redundant bounding boxes.

**4. ArUco Detection:** Similar to general bounding boxes, the ARUCO_BBOX_TRACKER_IOU_MATCH_THRESHOLD improves tracking association. The choice of KCF as the tracker might be better suited for the characteristics of ArUco tag tracking.

**5. Stanley Controller:** The significant reduction in the cross-track gain will make the rover less aggressive in correcting lateral errors. Decreasing the steering angle limit restricts the maximum steering input. The reduced prediction horizon and time step will result in the controller making decisions based on a much shorter-term prediction of the path, potentially making it more reactive to immediate path deviations but possibly less smooth in its overall trajectory.

**6. Stuck State:** Lowering the angular velocity threshold for detecting rotation makes the "stuck" detection more sensitive to even slight rotational movements.

**7. Global Handlers:** Introducing global handlers provides a single point of access to key autonomy components, potentially simplifying inter-module communication.

**8. Object Detection:** Since this is the majority of the PR, this section is longer. To summarize:
Refinement of Object Verification Logic: The VerifyingObjectState now has more robust logic for confirming the presence of a target object. It tracks the time an object has been consistently detected and its distance.

Handling of Verification Failure: If no object is detected for a certain duration or if the detected object is too far, the state triggers a "verify failed" event.

Handling of Verification Success: Once an object is consistently detected within a proximity threshold for a specified time, a "verify complete" event is triggered, potentially leading to the next state (e.g., ApproachingObjectState).

Integration with Global Handlers: The state interacts with globals::g_pObjectDetectionHandler to get object detections and globals::g_pStateMachineHandler to trigger events.

Waypoint Management: Upon successful verification and transition to eIdle, the next waypoint is popped from the queue, and saved states are cleared.

LoadDetectedObjects: This function aggregates detected objects from multiple object detectors. It uses futures to asynchronously request and retrieve detections.
IdentifyTargetObject: This function takes a vector of object detectors and identifies the "best" target object based on criteria like confidence, screen percentage, and age. It prioritizes Torch-detected objects and logs identified objects for debugging.

A new header file (ObjectDetectionUtility.hpp) defines a namespace (objectdetectutils) with structures and functions related to object detection.
Object struct: Defines a structure to hold information about a detected object, including its bounding box, confidence, distance, yaw angle, class name, detection time, detection method (e.g., Torch), image resolution, horizontal FOV, and geolocated position. It also includes overloaded equality and assignment operators.

### **Testing**
- Tested on Windows in the Dev Container. The code compiles and builds but object detection has not been testing in the simulator yet.

### **Additional Context**

No mallet Torch model added yet.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -36,7 +36,7 @@ set(CMAKE_SUPPRESS_DEVELOPER_WARNINGS 1 CACHE INTERNAL "Suppress developer warni
 ####################################################################################################################
 
 ## Enable or Disable Simulation Mode
-option(BUILD_SIM_MODE "Enable Simulation Mode" OFF)
+option(BUILD_SIM_MODE "Enable Simulation Mode" ON)
```
Comment (ClayJay3) [inline CMakeLists.txt]: Turn SIM mode off when merging to dev.

File: data/Cowsay_Cows/spider-family.cow [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    .--_'(  )'_--.   .'/()\'.   .'    '.  /(_)\  _\\()//_
+   / /` /`""`\ `\ \   \\  //   /   __   \       / //  \\ \
+    |  |  ><  |  |          ,  |   ><   |  ,     | \__/ |
+    \  \      /  /         . \  \      /  / .              _
+   _    '.__.'    _\(O)/_   \_'--`(  )'--'_/     __     _\(_)/_
+_\( )/_            /(_)\      .--'/()\'--.    | /  \ |   /(O)\
+ /(O)\  //  \\         _     /  /` '' `\  \  \_\\  //_/
+       _\\()//_     _\(_)/_    |        |      //()\\ 
+ jgs  / //  \\ \     /(o)\      \      /       \\  //
+       | \__/ |
```
Comment (ClayJay3) [inline data/Cowsay_Cows/spider-family.cow:19]: 🕷️

File: CMakeLists.txt (hunk 1 above)
Comment (three-halves) [inline CMakeLists.txt]: Is this sim mode supposed to be enabled?

File: src/states/ApproachingObjectState.cpp [hunk 3]
```diff
@@ -72,13 +89,172 @@ namespace statemachine
...
+            LOG_WARNING(logging::g_qSharedLogger,
+                        "ApproachingObjectState: Rover is too far from the original waypoint! Waypoint radius is {} meters, current distance is {} meters.",
+                        m_stGoalWaypoint.dRadius,
+                        stCurrentMeasurement.dDistanceMeters);
+            globals::g_pStateMachineHandler->HandleEvent(Event::eObjectUnseen);
+            return;
+        }
+
+        // Identify target object.
+        objectdetectutils::Object stBestObject;
```
Comment (ClayJay3) [inline src/states/ApproachingObjectState.cpp:123]: stBestObject is declared here, but never initialized. No call to the object detection handler is made to retrieve the objects, so this state won't work currently because it has no object data.

You need an identify object function similar to the tag state:
```
        // Identify target marker.
        tagdetectutils::ArucoTag stBestArucoTag, stBestTorchTag;
        statemachine::IdentifyTargetMarker(m_vTagDetectors, stBestArucoTag, stBestTorchTag, m_stGoalWaypoint.nID);
```

File: CMakeLists.txt (hunk 1 above)
Comment (frawgii) [reply CMakeLists.txt]: 'Tis not

File: src/algorithms/controllers/PredictiveStanleyController.cpp [hunk 4]
```diff
@@ -155,7 +156,7 @@ namespace controllers
             double dForwardVectorX = m_vReferencePath[m_nCurrentReferencePathTargetIndex + 1].GetUTMCoordinate().dEasting - stClosestWaypoint.GetUTMCoordinate().dEasting;
             double dForwardVectorY =
                 m_vReferencePath[m_nCurrentReferencePathTargetIndex + 1].GetUTMCoordinate().dNorthing - stClosestWaypoint.GetUTMCoordinate().dNorthing;
-            // Compute the norm and unit vector for the path segment.
+            // Compute the norm and unit vector for the path siSegment.
```
Comment (three-halves) [inline src/algorithms/controllers/PredictiveStanleyController.cpp]: What is an siSegment (as apposed to the previously named "segment")? I couldn't figure it out from looking through this file.

File: src/states/ApproachingObjectState.cpp [hunk 5]
```diff
@@ -72,13 +89,172 @@ namespace statemachine
...
+            if (stBestObject.stGeolocatedPosition.eType != geoops::WaypointType::eUNKNOWN)
+            {
+                // Calculate the geomeasurement to the object.
+                geoops::GeoMeasurement stObjectMeasurement =
+                    geoops::CalculateGeoMeasurement(stCurrentRoverPose.GetUTMCoordinate(), stBestObject.stGeolocatedPosition.GetUTMCoordinate());
+                // Update static variables.
+                dHeadingSetPoint = stObjectMeasurement.dStartRelativeBearing;
+                // Add the most recent geolocated object to the path plot.
+                m_pRoverPathPlot->AddPathPoint(stBestObject.stGeolocatedPosition.GetUTMCoordinate(), "DetectedObjects");
+            }
```
Comment (ClayJay3) [inline src/states/ApproachingObjectState.cpp]: Change the `AddPathPoint` to an `AddDot` call because we want the tag points to be plotted as dots instead of lines.

File: src/states/ApproachingObjectState.cpp [hunk 6]
```diff
@@ -72,13 +89,172 @@ namespace statemachine
...
+
+        //////////////////////////////////////////
+        // ---  Check if the rover is stuck --- //
+        //////////////////////////////////////////
+
+        // Check if stuck.
+        if (m_StuckDetector.CheckIfStuck(globals::g_pWaypointHandler->SmartRetrieveVelocity(), globals::g_pWaypointHandler->SmartRetrieveAngularVelocity()))
+        {
+            // Submit logger message.
+            LOG_NOTICE(logging::g_qSharedLogger, "NavigatingState: Rover has become stuck!");
```
Comment (ClayJay3) [inline src/states/ApproachingObjectState.cpp]: Change NavigatingState to ApproachingObjectState

File: src/states/ApproachingObjectState.cpp [hunk 7]
```diff
@@ -102,10 +278,27 @@ namespace statemachine
             {
                 // Submit logger message.
                 LOG_INFO(logging::g_qSharedLogger, "ApproachingObjectState: Handling ReachedObject event.");
-                // Pop old waypoint out of queue.
-                globals::g_pWaypointHandler->PopNextWaypoint();
-                // Change state.
-                eNextState = States::eIdle;
+
```
Comment (ClayJay3) [inline src/states/ApproachingObjectState.cpp:285]: eNextState should be our current state. So change eNextState to `States::eApproachingObject`

This is up at the top of the method where is says "Initialize member variables."

File: src/states/ApproachingObjectState.cpp [hunk 8]
```diff
@@ -102,10 +278,27 @@ namespace statemachine
...
+                }
+                else
+                {
+                    // Send multimedia command to update state display.
+                    globals::g_pMultimediaBoard->SendLightingState(MultimediaBoard::MultimediaBoardLightingState::eReachedGoal);
+                    // Pop old waypoint out of queue.
+                    globals::g_pWaypointHandler->PopNextWaypoint();
+                    // Clear saved search pattern state.
+                    globals::g_pStateMachineHandler->ClearSavedState(States::eApproachingObject);
+                    globals::g_pStateMachineHandler->ClearSavedState(States::eSearchPattern);
```
Comment (ClayJay3) [inline src/states/ApproachingObjectState.cpp]: Change to 
```
// Clear saved search pattern state.
globals::g_pStateMachineHandler->ClearSavedStates();
```
To make stuff easier and more future proof.

File: src/states/ApproachingObjectState.cpp [hunk 9]
```diff
@@ -24,14 +24,25 @@ namespace statemachine
...
+
+        // Store the state that got stuck and triggered an ObjectSeen event.
+        m_eTriggeringState = globals::g_pStateMachineHandler->GetPreviousState();
+
+        // Add the search and rover path layers to the plot.
+        m_pRoverPathPlot->CreateDotLayer("DetectedObjects", "orange");
+        m_pRoverPathPlot->CreateDotLayer("FinalTag", "green");
+        m_pRoverPathPlot->CreatePathLayer("RoverPath", "-.r*");
+
         m_nNumDetectionAttempts = 0;
```
Comment (ClayJay3) [inline src/states/ApproachingObjectState.cpp]: Remove this member variable, it is not used.

File: src/states/ApproachingObjectState.h [hunk 10]
```diff
@@ -25,14 +28,18 @@ namespace statemachine
...
     class ApproachingObjectState : public State
     {
         private:
-            int m_nNumDetectionAttempts;
+            statemachine::TimeIntervalBasedStuckDetector m_StuckDetector;
+            States m_eTriggeringState;
             bool m_bInitialized;
+            geoops::Waypoint m_stGoalWaypoint;
+            std::unique_ptr<logging::graphing::PathTracer> m_pRoverPathPlot;
+            int m_nNumDetectionAttempts;
```
Comment (ClayJay3) [inline src/states/ApproachingObjectState.h]: Remove this member variable, it is not used.

File: src/states/VerifyingObjectState.cpp [hunk 11]
```diff
@@ -120,6 +175,10 @@ namespace statemachine
                 globals::g_pMultimediaBoard->SendLightingState(MultimediaBoard::MultimediaBoardLightingState::eReachedGoal);
                 // Pop old waypoint out of queue.
                 globals::g_pWaypointHandler->PopNextWaypoint();
+                // Clear saved states.
+                globals::g_pStateMachineHandler->ClearSavedStates();
+                // Submit logger message.
+                LOG_NOTICE(logging::g_qSharedLogger, "VerifyingObjectState: Cleared old search pattern state and approaching object state from saved states.");
                 // Change state.
```
Comment (ClayJay3) [inline src/states/VerifyingObjectState.cpp]: Update log statement here to just say "VerifyingObjectState: Cleared old saved states."

**Review by ClayJay3 — CHANGES_REQUESTED** [review]
Small changes here and there. ApproachingObjectState is missing some code to actually get the tags, once that's fixed you're good. Looks great, good job!

File: src/algorithms/controllers/PredictiveStanleyController.cpp (hunk 4 above)
Comment (ClayJay3) [reply src/algorithms/controllers/PredictiveStanleyController.cpp]: bad find/replace. I'll fix it.

**Review by ClayJay3 — APPROVED** [review]
